src/Groups/matrices/linear_centralizer.jl

```diff
@@ -0,0 +1,400 @@
+   return res
+end
+
+_GL_order(n::Int, F::Ring) = _GL_order(n, order(F))
+
+function _SL_order(n::Int, q::fmpz)
+   res = q^div(n*(n-1),2)
+   for i in 2:n res *= (q^i-1) end
+   return res
+end
+
+_SL_order(n::Int, F::Ring) = _SL_order(n, order(F))
+
+########################################################################
+#
+# Centralizer in GL
+#
+########################################################################
+
+
+# returns as matrices
+function _gens_for_GL(n::Int, F::Ring)
+   n !=1 || return [matrix(F,1,1,[primitive_element(F)])]
+   if order(F)==2
+      h1 = identity_matrix(F,n)
+      h1[1,2] = 1
+      h2 = zero_matrix(F,n,n)
+      h2[1,n] = 1
+      for i in 1:n-1 h2[i+1,i] = -1 end
+      return h1,h2
+   else
+      h1 = identity_matrix(F,n)
+      h1[1,1] = primitive_element(F)
+      h2 = zero_matrix(F,n,n)
+      h2[1,1] = -1
+      h2[1,n] = 1
+      for i in 1:n-1 h2[i+1,i] = -1 end
+      return h1,h2
+   end
+end
+
+# returns as matrices
+# does the same as above with F = F[x]/(f), but every entry is replaced by a diagonal join of D corresponding blocks
+# ASSUMPTION: deg(f) > 1
+function _gens_for_GL_matrix(f::PolyElem, n::Int, F::Ring; D=1)
+   C = companion_matrix(f)
+   CP = evaluate(_centralizer(f),C)            # matrix of maximal order in the centralizer of the companion matrix
+
+   if n==1 return [diagonal_join([CP for i in 1:D])] end
+   h1 = identity_matrix(F,n*degree(f)*D)
+   insert_block!(h1,diagonal_join([CP for i in 1:D]),1,1)
+   h2 = zero_matrix(F,n*degree(f)*D,n*degree(f)*D)
+   for i in 1:(n-1)*degree(f)*D h2[i+degree(f)*D,i]=-1 end
+   for i in 1:degree(f)*D
+      h2[i,i]=-1
+      h2[i,i+(n-1)*degree(f)*D]=1
+   end
+   return h1,h2
+end
+
+
+# V = vector of integers of the dimensions of Jordan blocks
+# return the generators for the centralizers of the diagonal join of unipotent Jordan blocks of dimensions V
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+function _centr_unipotent(F::Ring, V::AbstractVector{Int}; isSL=false)
+   n = sum(V)
+   _lambda = gen(F)  # yes, gen(F) is correct; we don't need a primitive element in this case
+
+   # L = multiset(V)
+   L=[[V[1],1]]
+   for i in 2:length(V)
+      if V[i]==L[length(L)][1]
+         L[length(L)][2] += 1
+      else
+         push!(L, [V[i],1])
+      end
+   end
+   listgens = MatElem[]
+
+   # generators for GL + internal diagonal blocks
+   pos=1
+   for l in L
+      v_g = isSL ? _gens_for_SL(l[2],F) : _gens_for_GL(l[2],F)
+      for x in v_g
+         z = block_matrix(l[2],l[2],[x[i,j]*identity_matrix(F,l[1]) for i in 1:l[2] for j in 1:l[2]])
+         z = insert_block(identity_matrix(F,n),z,pos,pos)
+         push!(listgens,z)
+      end
+      if l[1]>1
+         for i in 1:l[1]-1
+         for j in 1:degree(F)
+            z = identity_matrix(F,l[1])
+            for k in 1:l[1]-i z[k,i+k]=_lambda^j end
+            z = insert_block(identity_matrix(F,n),z,pos,pos)
+            push!(listgens,z)
+         end
+         end
```

Comment:
You can combine nested for loops:
```suggestion
         for i in 1:l[1]-1, j in 1:degree(F)
            z = identity_matrix(F,l[1])
            for k in 1:l[1]-i z[k,i+k]=_lambda^j end
            z = insert_block(identity_matrix(F,n),z,pos,pos)
            push!(listgens,z)
         end
```